TelePharo-Core.package/TlpRemoteIDE.class/instance/connectTo..st

```diff
@@ -8,6 +8,7 @@ connectTo: aTCPAddress
 	remotePeer := connectingNetwork remotePeerAt: aTCPAddress.
 	remotePharo := TlpRemotePharo peer: remotePeer.
 	remotePharo registerTools: tools.
-	
+	remotePharo evaluate: [TlpRemotePharo resetCalypsoAnnouncers].
+	"When a new connection is made, the annoucers of Calypso on the server side need to be reseted, because if the network is broken, the server keeps the wrong old announcers"
```

Comment:
The cleanup should go into TlpRemoteUIManager>>#registerRemoteTool:as: which is executed on remote side:
```Smalltalk
registerRemoteTool: aSeamlessProxy as: aNameSelector
	| existingTool |
	self shouldBeRegistered.
	existingTool := Smalltalk tools remoteToolNamed: aNameSelector.
	existingTool isSeamlessProxy ifFalse: [ "otherwise new remote tools are connected"
		localTools at: aNameSelector ifAbsentPut: existingTool].
	...
```
We can do nice cleanup here without hardcoding Calypso. Something like:
```Smalltalk
existingTool isSeamlessProxy ifTrue: [
     (localTools at: aNameSelector) cleanupEnvironmentFromRemoteTool: existingTool
```
where ClyFullBrowser will do #resetCalypsoAnnouncer (more comments below)

Reply:
![image](https://user-images.githubusercontent.com/39618015/75975139-42823d80-5ed8-11ea-8419-6a4d6db634ce.png)

This "existingTool" is a ClyFullBrowser, not a SeamlessProxy, so it is not triggering the cleanup method. 

Why is checking if exisitingTool is a SeamlessProxy or no? When it is and when it is not?

Reply:
This method is called 2 times, the first one the existing tool is ClyFullBrowser and the second one is GTGenericStackDebugger 

![image](https://user-images.githubusercontent.com/39618015/76005861-e9340180-5f0b-11ea-848c-a5dfc76876d2.png)

Reply:
This method is called 2 times, the first one the existingTool variable has the value ClyFullBrowser and the second one has GTGenericStackDebugger.

Can we use something like that:

```
(existingTool isKindOf: ClyFullBrowser class) ifTrue: [
		(localTools at: aNameSelector) cleanupEnvironmentFromRemoteTool: existingTool ].
``` 
I used and it works well. 

![image](https://user-images.githubusercontent.com/39618015/76006003-21d3db00-5f0c-11ea-99f7-92519c3a3c31.png)

Reply:
Humm.. this cleanup it's not working. It removes the queryCache, but this some times doesn't make the browser work again on the client-side

![image](https://user-images.githubusercontent.com/39618015/76009478-8e9da400-5f11-11ea-8d70-4a0866259a36.png)

Reply:
> This method is called 2 times, the first one the existingTool variable has the value ClyFullBrowser and the second one has GTGenericStackDebugger.

Yes, the cleanup method needs to be implemented in all tool classes which TelePharo installs on remote image

Reply:
> This "existingTool" is a ClyFullBrowser, not a SeamlessProxy, so it is not triggering the cleanup method.
> 
> Why is checking if exisitingTool is a SeamlessProxy or no? When it is true and when it is not?

When existingTool is not a proxy it means that we are connecting first time to the image and therefore nothing to clean. When we get a proxy it is from previous "broken session" which requires cleanup. 

(Notice there is also non "broken" scenario when we connecting to the image which has live TelePharo session from some other client (two devs could connect to remote image). In current logic the last will preempt the first. But full cleanup from the first client is missing)

Reply:
> When existingTool is not a proxy it means that we are connecting first time to the image and therefore nothing to clean. When we get a proxy it is from previous "broken session" which requires cleanup.


I did more tests using "existingTool isSeamlessProxy ifTrue:" , but it is never true. In this way, the #cleanupEnvironmentFromRemoteTool: is never called. 

Another thing, I think we don't need pass parameters to the cleanup method, because we are not using it, like in #cleanupEnvironmentFromRemoteTool: aRemoteBrowser. This aRemoteBrowser is not used.

Reply:
> >
> I did more tests using "existingTool isSeamlessProxy ifTrue:" , but it is never true. In this way, the #cleanupEnvironmentFromRemoteTool: is never called.
> 
> Another thing, I think we don't need pass parameters to the cleanup method, because we are not using it, like in #cleanupEnvironmentFromRemoteTool: aRemoteBrowser. This aRemoteBrowser is not used.

ping...

@dionisiydk this logic of "existingTool isSeamlessProxy" is not matching with you explained.

Reply:
> > > 
> > 
> > 
> > I did more tests using "existingTool isSeamlessProxy ifTrue:" , but it is never true. In this way, the #cleanupEnvironmentFromRemoteTool: is never called.
> > Another thing, I think we don't need pass parameters to the cleanup method, because we are not using it, like in #cleanupEnvironmentFromRemoteTool: aRemoteBrowser. This aRemoteBrowser is not used.
> 
> ping...
> 
> @dionisiydk this logic of "existingTool isSeamlessProxy" is not matching with you explained, maybe I didn't get. But isSeamlessProxy is not true in these conditions that you described.

Hm.. When I connect at first time the tool is proxy - script on remote side returns true:  
```Smalltalk
(Smalltalk tools remoteToolNamed: #browser) isSeamlessProxy "==>true"
```
When I disconnect (remotePharo disconnect) it returns false

Reply:
In problem scenario we do not perform disconnect and therefore the proxy is staying in the tools.
But anyway we can do cleanup without this condition, just every time when we perform connection.

Reply:
> ```smalltalk
> ```smalltalk
> (Smalltalk tools remoteToolNamed: #browser) isSeamlessProxy "==>true"
> ```
> ```

I get the point. You run this after performing this part of code:
> ```smalltalk
> ```smalltalk
> Smalltalk tools registerRemoteTool: aSeamlessProxy as: aNameSelector.> ```
> ```
> ```

So I put the ifTrue after this part and checking directly at "Smalltalk tools", because the temporary variable existingTool is old at this point.

And I put the method #cleanupEnvironmentFromRemoteTool in this class without any parameter.

Reply:
> In problem scenario we do not perform disconnect and therefore the proxy is staying in the tools.

In this problem scenario, we perform disconnect. The other problem scenario when we don't perform disconnect, is when having the timeout issue. 

> But anyway we can do cleanup without this condition, just every time when we perform connection.

What do you think about the cleanup that I committed now?